windows: expose support for recursive directory watches

nicks opened the pull request:
#### What does this pull request do?
On Windows, ReadDirectoryChanges allows recursive watches, so I added a function call to enable it

I saw some discussion on this in these issues: 
https://github.com/fsnotify/fsnotify/issues/18
https://github.com/fsnotify/fsnotify/issues/21

but given that there's not much activity in this repo, it didn't seem worth blocking on a user-space solutoin to expose this

#### How should this be manually tested?
I've run the unittest on Windows and confirms it watches things successfully. There's also a new test in integration_test, adapted from the existing subdirectory test.

cobolbaby commented:
LGTM.

@houndci-bot , could merge this PR?

@nicks  Another question is, after configuring the `recursive` option of `syscall.ReadDirectoryChanges`, will it hold a lot of memory like the recursive directory `addWatch`?

The problem I currently encounter is that due to the limitations of business scenarios, there will be a lot of monitoring subdirectories, and the recursive `addWatch` operation of a directory will occupy more than 1G of memory space, which obviously takes up a lot of resources.

hu13 commented:
@nicksantos if this gets merged, you may need to include this https://github.com/fsnotify/fsnotify/pull/361 when your recursive watch hits a filepath that exceeds windows' syscall.MAX_PATH.

mattn commented on fen.go:
```diff
@@ -21,6 +22,13 @@ func NewWatcher() (*Watcher, error) {
+// if this file system supports that.
+// Must be called before Add/Remove.
+func (w *Watcher) SetRecursive() error {
+	return fmt.Errorf("Not supported")
```
Please use errors.New

mattn commented on inotify.go:
```diff
@@ -69,6 +69,13 @@ func (w *Watcher) isClosed() bool {
+// if this file system supports that.
+// Must be called before Add/Remove.
+func (w *Watcher) SetRecursive() error {
+	return fmt.Errorf("Not supported")
```
ditto

mattn commented on integration_test.go:
```diff
@@ -1226,6 +1226,105 @@ func TestRemoveWithClose(t *testing.T) {
+	defer os.RemoveAll(testDir)
+
+	testSubDir := filepath.Join(testDir, "sub")
+	testSubDirFile := filepath.Join(testDir, "sub/TestFsnotifyFile1.testfile")
```
```
filepath.Join(testDir, "sub", "TestFsnotifyFile1.testfile")
```

mattn replied:
Or
```
testSubDirFile := filepath.Join(testSubDir, "TestFsnotifyFile1.testfile")
```

mattn commented on integration_test.go:
```diff
@@ -1226,6 +1226,105 @@ func TestRemoveWithClose(t *testing.T) {
+				expected = expected || event.Name == filepath.Clean(testSubDirFile)
+			}
+			if expected {
+				t.Logf("event received: %s", event)
```
Please use `%v` instead of `%s`

mattn commented on integration_test.go:
```diff
@@ -1226,6 +1226,105 @@ func TestRemoveWithClose(t *testing.T) {
+					deleteReceived.increment()
+				}
+			} else {
+				t.Logf("unexpected event received: %s", event)
```
ditto

mattn commented on integration_test.go:
```diff
@@ -1226,6 +1226,105 @@ func TestRemoveWithClose(t *testing.T) {
+
+	// Create sub-directory
+	if err := os.Mkdir(testSubDir, 0777); err != nil {
+		t.Fatalf("failed to create test sub-directory: %s", err)
```
Use `%v`

mattn commented on integration_test.go:
```diff
@@ -1226,6 +1226,105 @@ func TestRemoveWithClose(t *testing.T) {
+	var fs *os.File
+	fs, err = os.OpenFile(testSubDirFile, os.O_WRONLY|os.O_CREATE, 0666)
+	if err != nil {
+		t.Fatalf("creating test file failed: %s", err)
```
ditto

mattn commented on kqueue.go:
```diff
@@ -63,6 +63,13 @@ func NewWatcher() (*Watcher, error) {
+// if this file system supports that.
+// Must be called before Add/Remove.
+func (w *Watcher) SetRecursive() error {
+	return fmt.Errorf("Not supported")
```
Use errors.New

nathany commented:
@mattn Thanks for doing the review. Since this PR is nearly 2 years old, if @nicks doesn't respond, it may be necessary to `gh pr checkout` this branch and open a new pull request with the suggested fixes.

nathany commented:
#397

nicks commented:
thanks! addressed all comments. most of the comments were on to changed things that were also issues in other tests, so i took the liberty of addressing those too. also rebased on main.

nathany commented on windows.go:
```diff
@@ -37,17 +38,29 @@ func NewWatcher() (*Watcher, error) {
 	return w, nil
 }
 
+// SetRecursive makes future Add() calls watch directories recursively
+// if this file system supports that.
+// Must be called before Add/Remove.
+func (w *Watcher) SetRecursive() error {
```
I'm not sure about this API -- what do others think?

Adding a separate API is good in that there aren't any breaking changes for existing users.

There is no way to disable the recursive flag. Is that important?

nicks replied:
ya, i could go either way on this.

my thinking was that you want the API to be as restrictive as possible - no switching back and forth between recursive and non-recursive, no switching modes once the watcher has started listening on events.

But certainly `SetRecursive(bool)` would be a lot more idiomatic.

nathany replied:
There are a lot of different options, and it's important to get the API right, because the same API would need to support a user-space recursive watcher (#18) and [FSEvents](https://github.com/fsnotify/fsevents) on macOS.

@pbnjay @rogpeppe Having look at the Mac side, do you have any opinions on how this API should work?

If I recall, macOS is always recursive when making a watch?

This option seems pretty good and is backwards compatible -- a method on the watcher.

Other options would be some other form of NewWatcher with configuration. If we want the watcher to be 100% recursive or 100% not, this may be better.

Or having a separate `AddRecursive` function, which would allow intermixing of recursive and not-recursive. But we may wind up with too many variants -- like with and without Symlinks, Recursive, Ops filtering, etc.

hu13 replied:
@nathany 

I can confirm that fsevents on macOS is recursive by default.

I am in support of `Other options would be some other form of NewWatcher with configuration. If we want the watcher to be 100% recursive or 100% not, this may be better`.

I don't see a concrete use case in which we want to keep switching back and forth. Most of time, people want a recursive watch. If there are paths that they are not interested in, they can reasonably add additional filtering-out step at their application level to ignore such events.

Please let us know what else to do for this to be acceptable. thanks

hu13 replied:
As a side note, we forked this and enabled a recursive watch like this in our production software and we don't observe issues https://github.com/fsnotify/fsnotify/issues/259 and https://github.com/fsnotify/fsnotify/issues/243.

nathany replied:
I think we could just go with this API for now. 

It's possible to change it later in a v2 of the library.

We probably just need someone to update the branch, merge it, and do the release

pbnjay replied:
(Sorry this somehow skipped my inbox)

Yes macOS is recursive by default, but it could be a challenge to implement logic to "disable" that however. It sounds like could be the default fsnotify semantics?

arp242 replied:
One possibility would be to use `Add("dir/**")` or `Add("dir/...")`.

Especially since recursive operations may be expensive on some platforms, there may be use cases where you might want things like "recursively watch ~/.cache/go-build, and non-recursively watch ~/.cache/go-mod", or something like that.

The advantage of using a string like that is that #352 can be added without too much friction too as e.g. `AddMask("path", fsnotify.CREATE | fsnotify.DELETE)` or some such.

nathany replied:
Would it make sense to have a syntax like Glob?

https://pkg.go.dev/path/filepath@go1.18.4#Glob
https://pkg.go.dev/io/fs@go1.18.4#Glob

nathany replied:
One challenge with this approach is what happens on BSD (kqueue) or Linux (inotify), since only Windows and FSEvents on macOS support recursive watching. 

Is it an error on some platforms? At least until such time as a user-space recursive watcher (#18) covers those platforms?

arp242 replied:
> One challenge with this approach is what happens on BSD (kqueue) or Linux (inotify), since only Windows and FSEvents on macOS support recursive watching.

Yeah, .Add() can return an error in those cases, similar to how SetRecursive() does now. Something like:


	func recursive(path string) (string, bool) {
		if strings.HasSuffix(path, "/...") || strings.HasSuffix(path, `\...`) {
			return path[:len(path)-4], true
		}
		return path, false
	}

	path, recurse := recursive(path)
	if recurse {
    	return errors.New("recursion not supported on this system")
	}

Or for Windows and other systems where it works:

    recurse := false
    path, recurse = recursive(path)


> Would it make sense to have a syntax like Glob?

I'd be in favour of that; there's an old issue for that (#10).

Using the `/...` syntax is probably better, since e.g. something like `**/*.zip` won't work and I don't think we need to support that (it'll be tricky to get that to work well for all cases).

nathany replied:
Yeah, that makes sense, and there is precedence for that format (e.g. `go test ./...`). Sounds good.

arp242 commented:
There's still quite a bit of work that needs to be done in terms of API; not 100% sure what it will look like yet, but certainly different from this patch. For example #521 adds a very different way of passing options, and I probably want to use `path/...` anyway rather than options.

The core is pretty trivial (just pass a boolean around), just need to figure out an API that works well. Either way, much of this will have to be re-done, and I don't think there's much value in keeping this PR open or basing it this PR, as most things will be different.

Can subscribe to #21 for notification when it's added.
